### PR TITLE
[NAE-1945] External resource loader

### PR DESCRIPTION
**Pull request by Kovy95**
# Description

- implementing our own resource loader for files from directory reources/ in root

Implements [NAE-1945]

## Dependencies

none

### Third party dependencies

No new dependencies were introduced

### Blocking Pull requests

There are no dependencies on other PR

## How Has Been This Tested?

manually

### Test Configuration

| Name                | Tested on |
|---------------------| --------- |
| OS                  |     LinuxMint 20     |
| Runtime             |      Java 11     |
| Dependency Manager  |    	Maven 3.8.4       |
| Framework version   |     SpringBoot 2.6.2      |
| Run parameters      |           |
| Other configuration |           |

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My changes have been checked, personally or remotely, with @...
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have resolved all conflicts with the target branch of the PR
- [ ] I have updated and synced my code with the target branch
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes:
    - [ ] Lint test
    - [ ] Unit tests
    - [ ] Integration tests
- [ ] I have checked my contribution with code analysis tools:
    - [ ] [SonarCloud](https://sonarcloud.io/project/overview?id=netgrif_application-engine)
    - [ ] [Snyk](https://app.snyk.io/org/netgrif/project/ea82b3b8-658d-41f5-89a7-76cd600da01f)
- [ ] I have made corresponding changes to the documentation:
    - [ ] Developer documentation
    - [ ] User Guides
    - [ ] Migration Guides

[NAE-1945]: https://netgrif.atlassian.net/browse/NAE-1945?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/com/netgrif/application/engine/resource/domain/NetgrifResourceLoader.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    public NetgrifResourceLoader(ResourceLoader delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public Resource getResource(String location) {
+        if(location.startsWith(RESOURCE_PREFIX)) {
+            String path = location.substring(RESOURCE_PREFIX.length());
+            NetgrifResource resource = new NetgrifResource();
+            return resource.getResource("resources/" + path);
```
Comment (tuplle) [inline src/main/java/com/netgrif/application/engine/resource/domain/NetgrifResourceLoader.java:21]: extract path prefix as a constant

File: src/main/java/com/netgrif/application/engine/resource/service/NetgrifResourceLoaderProcessor.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import com.netgrif.application.engine.resource.domain.NetgrifResourceLoader;
+import org.springframework.context.ResourceLoaderAware;
+import org.springframework.core.io.DefaultResourceLoader;
+import org.springframework.core.io.ProtocolResolver;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.ResourceLoader;
+import org.springframework.stereotype.Component;
+
+@Component
+public class NetgrifResourceLoaderProcessor implements ResourceLoaderAware, ProtocolResolver {
```
Comment (tuplle) [inline src/main/java/com/netgrif/application/engine/resource/service/NetgrifResourceLoaderProcessor.java:12]: Rename class to `ExternalResourceLoaderProcessor`

File: src/main/java/com/netgrif/application/engine/resource/service/NetgrifResourceLoaderProcessor.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    @Override
+    public void setResourceLoader(ResourceLoader resourceLoader) {
+        if(DefaultResourceLoader.class.isAssignableFrom(resourceLoader.getClass())) {
+            ((DefaultResourceLoader)resourceLoader).addProtocolResolver(this);
+        } else {
+            System.out.println("Could not assign protocol loader.");
```
Comment (tuplle) [inline src/main/java/com/netgrif/application/engine/resource/service/NetgrifResourceLoaderProcessor.java:19]: Change it to `log.error`

**Review by tuplle — APPROVED** [review]

File: docs/resources/resources_loading.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# Resource Loading
+
+If you want to load resources, which will not be in JAR(for example big files), you can use our resource loader.
+ResourceLoader return InputStreamResource, from which you can use InputStream, and loading resources from directory **resource/** in root of the project.
```
[the remark this answers is not in the document]
Comment (Kovy95) [reply docs/resources/resources_loading.md]: done

File: src/main/java/com/netgrif/application/engine/resource/domain/NetgrifResource.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+import org.springframework.core.io.InputStreamResource;
+import org.springframework.core.io.Resource;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+public class NetgrifResource {
```
[the remark this answers is not in the document]
Comment (Kovy95) [reply src/main/java/com/netgrif/application/engine/resource/domain/NetgrifResource.java]: done

File: src/main/java/com/netgrif/application/engine/resource/domain/NetgrifResourceLoader.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package com.netgrif.application.engine.resource.domain;
+
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.ResourceLoader;
+
+public class NetgrifResourceLoader implements ResourceLoader {
+
+    public static final String RESOURCE_PREFIX = "resource://";
```
[the remark this answers is not in the document]
Comment (Kovy95) [reply src/main/java/com/netgrif/application/engine/resource/domain/NetgrifResourceLoader.java]: done

[the remark this answers is not in the document]
Comment (Kovy95) [reply src/main/java/com/netgrif/application/engine/resource/domain/NetgrifResourceLoader.java]: done

File: src/main/java/com/netgrif/application/engine/resource/service/NetgrifResourceLoaderProcessor.java (hunk 2 above)
Comment (Kovy95) [reply src/main/java/com/netgrif/application/engine/resource/service/NetgrifResourceLoaderProcessor.java]: done

File: src/main/java/com/netgrif/application/engine/resource/service/ExternalResourceLoaderProcessor.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        if(DefaultResourceLoader.class.isAssignableFrom(resourceLoader.getClass())) {
+            ((DefaultResourceLoader)resourceLoader).addProtocolResolver(this);
+        } else {
+            log.error("Could not assign protocol loader.");
+        }
+    }
+
+    @Override
+    public Resource resolve(String location, ResourceLoader resourceLoader) {
+        if(location.startsWith(ExternalResourceLoader.RESOURCE_PREFIX) ||
```
Comment (renczesstefan) [inline src/main/java/com/netgrif/application/engine/resource/service/ExternalResourceLoaderProcessor.java:27]: For better readability, may export this long if condition to a function.

**Review by renczesstefan — APPROVED** [review]

**Review by renczesstefan — APPROVED** [review]

File: src/main/java/com/netgrif/application/engine/resource/domain/ExternalResourceLoader.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Override
+    public ClassLoader getClassLoader() {
+        return this.delegate.getClassLoader();
+    }
+
+    private Resource substringAndGetResource(String location, String prefix) {
+        String path = location.substring(prefix.length());
+        ExternalResource resource = new ExternalResource();
+        return resource.getResource(StringUtils.chop(prefix)+ "/" + path);
```
Comment (minop) [inline src/main/java/com/netgrif/application/engine/resource/domain/ExternalResourceLoader.java:45]: shouldn't `"/"` be `File.pathSeparator` instead?

File: docs/resources/resources_loading.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
+# Resource Loading
+
+If you want to load resources, which will not be in JAR(for example big files), you can use our resource loader.
+ResourceLoader return InputStreamResource, from which you can use InputStream, and loading resources from directory **resource/**  inworking directory of the app.
+The prefix for NetgrifResourceLoaser is
```
[the remark this answers is not in the document]
Comment (tuplle) [reply docs/resources/resources_loading.md:5]: Corrected

File: src/main/java/com/netgrif/application/engine/resource/domain/ExternalResourceLoader.java (hunk 8 above)
Comment (tuplle) [reply src/main/java/com/netgrif/application/engine/resource/domain/ExternalResourceLoader.java:45]: Done

File: src/main/java/com/netgrif/application/engine/resource/service/ExternalResourceLoaderProcessor.java (hunk 7 above)
Comment (tuplle) [reply src/main/java/com/netgrif/application/engine/resource/service/ExternalResourceLoaderProcessor.java:27]: Resolved

**Review by renczesstefan — APPROVED** [review]

**Review by tuplle — APPROVED** [review]

**Review by machacjozef — APPROVED** [review]

**Pull request merged**